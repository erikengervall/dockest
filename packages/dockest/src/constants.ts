export const LOG_LEVEL = {
  NOTHING: 0,
  ERROR: 1,
  WARN: 2,
  INFO: 3,
  DEBUG: 4,
}

export const GENERATED_COMPOSE_FILE_PATH = `${process.cwd()}/docker-compose.dockest-generated.yml`

export const DOCKEST_ATTACH_TO_PROCESS = 'DOCKEST_ATTACH_TO_PROCESS'

export const BRIDGE_NETWORK_NAME = `dockest_bridge_network`

export const DOCKEST_HOST_ADDRESS = 'host.dockest-runner.internal'
export const DEFAULT_HOST_NAME = 'host.docker.internal'

export const DEFAULT_OPTS = {
  composeFile: 'docker-compose.yml',
  jestOpts: {
    projects: ['.'],
    runInBand: true,
  },
  logLevel: LOG_LEVEL.INFO,
}

export const DEFAULT_$ = {
  dockestServices: [],
  hostname: process.env.HOSTNAME || DEFAULT_HOST_NAME,
  isInsideDockerContainer: false,
  jestRanWithResult: false,
  perfStart: 0,
  runners: [],
}

/**
 * Released 2017-05-06: https://github.com/facebook/jest/releases/tag/v20.0.0
 */
export const MINIMUM_JEST_VERSION = '20.0.0'
