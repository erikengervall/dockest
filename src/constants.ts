export const COLORS = {
  MISC: {
    BLINK: '\x1b[5m',
    BRIGHT: '\x1b[1m',
    DIM: '\x1b[2m',
    HIDDEN: '\x1b[8m',
    RESET: '\x1b[0m',
    REVERSE: '\x1b[7m',
    UNDERSCORE: '\x1b[4m',
  },
  FG: {
    BLACK: '\x1b[30m',
    BLUE: '\x1b[34m',
    CYAN: '\x1b[36m',
    GREEN: '\x1b[32m',
    MAGENTA: '\x1b[35m',
    RED: '\x1b[31m',
    WHITE: '\x1b[37m',
    YELLOW: '\x1b[33m',
  },
  BG: {
    BLACK: '\x1b[40m',
    BLUE: '\x1b[44m',
    CYAN: '\x1b[46m',
    GREEN: '\x1b[42m',
    MAGENTA: '\x1b[45m',
    RED: '\x1b[41m',
    WHITE: '\x1b[47m',
    YELLOW: '\x1b[43m',
  },
}

export const LOG_SYMBOLS: readonly string[] = [
  'üêâ ',
  'üêí ',
  'üêô ',
  'üêû ',
  'üê• ',
  'üêº ',
  'üêø ',
  'ü¶Ç ',
  'ü¶É ',
  'ü¶Ñ ',
  'ü¶ä ',
  'ü¶ã ',
  'ü¶ç ',
  'ü¶ñ ',
  'ü¶ö ',
]

export const LOG_LEVEL = {
  NOTHING: 0,
  ERROR: 1,
  WARN: 2,
  INFO: 3,
  DEBUG: 4,
}

export const DEFAULT_USER_CONFIG = {
  afterSetupSleep: 0,
  dev: { debug: false },
  dumpErrors: false,
  exitHandler: null,
  logLevel: LOG_LEVEL.INFO,
  runInBand: true,
  composeFile: [],
  guessRunnerType: false,
}

export const INTERNAL_CONFIG = {
  failedTeardowns: [],
  jestRanWithResult: false,
  perfStart: 0,
  isInsideDockerContainer: false,
}

export const PROCESS_TEST_ENV = 'dockest-test'

export const GENERATED_COMPOSE_FILE_NAME = 'docker-compose.dockest-generated.yml'
export const GENERATED_RUNNER_COMPOSE_FILE_NAME = 'docker-compose.dockest-generated-runner.yml'
export const GENERATED_COMPOSE_FILE_PATH = `${process.cwd()}/${GENERATED_COMPOSE_FILE_NAME}`
export const GENERATED_RUNNER_COMPOSE_FILE_PATH = `${process.cwd()}/${GENERATED_RUNNER_COMPOSE_FILE_NAME}`

export const BRIDGE_NETWORK_NAME = `dockest_bridge_network`
