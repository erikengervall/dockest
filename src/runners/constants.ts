import { SharedDefaultableConfigProps } from './@types'

export const SHARED_DEFAULT_CONFIG_PROPS: SharedDefaultableConfigProps = {
  build: '',
  commands: [],
  connectionTimeout: 5,
  dependsOn: [],
  host: 'localhost',
  image: undefined,
  networks: undefined,
  ports: [],
  props: {},
  responsivenessTimeout: 30,
}
