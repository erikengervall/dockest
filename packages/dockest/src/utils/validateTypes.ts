import { COLORS } from '../constants'

const { keys, values } = Object

type isType = (_?: any) => boolean

const {
  FG: { RED, GREEN },
  MISC: { RESET },
} = COLORS

const getExpected = (typeValidator: any): string =>
  typeValidator.expected || typeValidator.name.substring(2).toLowerCase()

const getReceived = (value: any): string => (isArray(value) ? `${typeof value[0]}[]` : `${value} (${typeof value})`)

const validateTypes = (schema?: { [key: string]: isType }, config?: { [key: string]: any }): string[] => {
  const failures: string[] = []

  if (!schema) {
    failures.push(`${RED}Missing validation schema${RESET}`)
    return failures
  }

  if (!config) {
    failures.push(`${RED}Missing config to validate${RESET}`)
    return failures
  }

  keys(schema).forEach(schemaKey => {
    const value = config[schemaKey]

    if (value) {
      const typeValidatorFn = schema[schemaKey]

      // TODO: Allow sending multiple things
      // if (isArray(typeValidatorFn)) {
      // }

      if (typeValidatorFn(value) === false) {
        const expected = `${RED}Expected${RESET} ${getExpected(typeValidatorFn)}`
        const received = `${GREEN}Received${RESET} ${getReceived(value)}`

        failures.push(`${schemaKey}${RESET}: ${expected} | ${received}`)
      }
    } else {
      failures.push(`${RED}${schemaKey}${RESET}: Schema-key missing in config`)
    }
  })

  return failures
}

const isString: isType = _ => (_ && typeof _ === 'string') || _ instanceof String
const isNumber: isType = _ => _ && typeof _ === 'number' && isFinite(_)
const isArray = (_?: any): boolean => _ && typeof _ === 'object' && _.constructor === Array
const isArrayOfType = (fn: isType) => {
  const isArrayOfType = (_?: any): boolean => isArray(_) && !_.some((_?: any) => !fn(_))
  isArrayOfType.expected = `${fn.name.substring(2).toLowerCase()}[]`
  return isArrayOfType
}
const isFunction: isType = _ => _ && typeof _ === 'function'
const isObject: isType = _ => _ && typeof _ === 'object' && _.constructor === Object
const isObjectWithValuesOfType = (fn: isType) => {
  const isObjectWithValuesOfType = (_?: any): boolean => isObject(_) && !values(_).some((_?: any) => !fn(_))
  isObjectWithValuesOfType.expected = `{ [prop: string]: ${fn.name.substring(2).toLowerCase()} }`
  return isObjectWithValuesOfType
}
const isNull: isType = _ => _ === null
const isUndefined: isType = _ => typeof _ === 'undefined'
const isBoolean: isType = _ => _ && typeof _ === 'boolean'
const isRegExp: isType = _ => _ && typeof _ === 'object' && _.constructor === RegExp
const isError: isType = _ => _ && _ instanceof Error && typeof _.message !== 'undefined'
const isDate: isType = _ => _ && _ instanceof Date
const isSymbol: isType = _ => _ && typeof _ === 'symbol'
const isAny: isType = _ => _ && !isNull(_) && !isUndefined(_)
const isOneOf = (haystack: any[]) => {
  const isOneOf = (needle: any): boolean => isArray(haystack) && !!haystack.find(_ => _ === needle)
  isOneOf.expected = `oneOf [${haystack.join(', ')}]`
  return isOneOf
}
const OR = 'OR'
const AND = 'AND'

validateTypes.isString = isString
validateTypes.isNumber = isNumber
validateTypes.isArray = isArray
validateTypes.isArrayOfType = isArrayOfType
validateTypes.isFunction = isFunction
validateTypes.isObject = isObject
validateTypes.isObjectWithValuesOfType = isObjectWithValuesOfType
validateTypes.isNull = isNull
validateTypes.isUndefined = isUndefined
validateTypes.isBoolean = isBoolean
validateTypes.isRegExp = isRegExp
validateTypes.isError = isError
validateTypes.isDate = isDate
validateTypes.isSymbol = isSymbol
validateTypes.isAny = isAny
validateTypes.isOneOf = isOneOf
validateTypes.OR = OR
validateTypes.AND = AND

export default validateTypes
