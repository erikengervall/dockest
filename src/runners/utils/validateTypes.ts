import { COLORS } from '../../constants'

const {
  FG: { RED, GREEN },
  MISC: { RESET },
} = COLORS

interface OObject {
  [key: string]: any
}
type isType = (_?: any) => boolean

const getExpected = (typeValidator: any): string =>
  typeValidator.expected || typeValidator.name.substring(2).toLowerCase()

const getReceived = (value: any): string =>
  isArray(value) ? `${typeof value[0]}[]` : `${value} (${typeof value})`

const validateTypes = (schema: OObject, config?: OObject): string[] => {
  if (!config) {
    return [`${RED}No config found${RESET}`]
  }

  const failures: string[] = []

  Object.keys(schema).forEach(schemaKey => {
    const value = config[schemaKey]

    if (value) {
      const typeValidator = schema[schemaKey]

      if (isArray(typeValidator)) {
        // Allow sending multiple thangz
      }

      if (!typeValidator(value)) {
        const testedSchemaKey = `${schemaKey}${RESET}`
        const expected = `${RED}Expected${RESET} ${getExpected(typeValidator)}`
        const received = `${GREEN}Received${RESET} ${getReceived(value)}`

        failures.push(`${testedSchemaKey}: ${expected} | ${received}`)
      }
    } else {
      failures.push(`${RED}${schemaKey}${RESET}: Schema-key missing in config`)
    }
  })

  return failures
}

const isString: isType = _ => (_ && typeof _ === 'string') || _ instanceof String
const isNumber: isType = _ => _ && typeof _ === 'number' && isFinite(_)
const isArray = (_?: any): boolean => _ && typeof _ === 'object' && _.constructor === Array
const isArrayOfType = (fn: isType) => {
  const isArrayOfType = (_?: any): boolean => isArray(_) && !_.some((_?: any) => !fn(_))
  isArrayOfType.expected = `${fn.name.substring(2).toLowerCase()}[]`
  return isArrayOfType
}
const isFunction: isType = _ => _ && typeof _ === 'function'
const isObject: isType = _ => _ && typeof _ === 'object' && _.constructor === Object
const isObjectWithValuesOfType = (fn: isType) => {
  const isObjectWithValuesOfType = (_?: any): boolean =>
    isObject(_) && !Object.values(_).some((_?: any) => !fn(_))
  isObjectWithValuesOfType.expected = `{ [prop: string]: ${fn.name.substring(2).toLowerCase()} }`
  return isObjectWithValuesOfType
}
const isNull: isType = _ => _ === null
const isUndefined: isType = _ => typeof _ === 'undefined'
const isBoolean: isType = _ => _ && typeof _ === 'boolean'
const isRegExp: isType = _ => _ && typeof _ === 'object' && _.constructor === RegExp
const isError: isType = _ => _ && _ instanceof Error && typeof _.message !== 'undefined'
const isDate: isType = _ => _ && _ instanceof Date
const isSymbol: isType = _ => _ && typeof _ === 'symbol'
const isAny: isType = _ => _ && !isNull(_) && !isUndefined(_)
const isOneOf = (haystack: any[]) => {
  const isOneOf = (needle: any): boolean => isArray(haystack) && !!haystack.find(_ => _ === needle)
  isOneOf.expected = `oneOf [${haystack.join(', ')}]`
  return isOneOf
}
const OR = 'OR'
const AND = 'AND'

validateTypes.isString = isString
validateTypes.isNumber = isNumber
validateTypes.isArray = isArray
validateTypes.isArrayOfType = isArrayOfType
validateTypes.isFunction = isFunction
validateTypes.isObject = isObject
validateTypes.isObjectWithValuesOfType = isObjectWithValuesOfType
validateTypes.isNull = isNull
validateTypes.isUndefined = isUndefined
validateTypes.isBoolean = isBoolean
validateTypes.isRegExp = isRegExp
validateTypes.isError = isError
validateTypes.isDate = isDate
validateTypes.isSymbol = isSymbol
validateTypes.isAny = isAny
validateTypes.isOneOf = isOneOf
validateTypes.OR = OR
validateTypes.AND = AND
export default validateTypes
