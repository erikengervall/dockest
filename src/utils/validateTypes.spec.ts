import validateTypes from './validateTypes'
import { LOG_LEVEL } from '../constants'

const { values } = Object

describe('validateTypes', () => {
  let config: { [key: string]: any }

  beforeEach(() => {
    config = {
      service: 'username',
      commands: ['first command', 'second command'],
      variousStuff: ['first command', 1, () => ({})],
      port: 1337,
      ports: ['1337:1337', '1338:1338'],
      autoCreateTopics: true,
      portMapping: { '1337': '1338' },
      logLevel: 1,
    }
  })

  describe('happy', () => {
    it('should work for valid config', () => {
      const validationSchema = {
        service: validateTypes.isString,
        commands: validateTypes.isArrayOfType(validateTypes.isString),
        variousStuff: validateTypes.isArrayOfType(validateTypes.isAny),
        port: validateTypes.isNumber,
        ports: validateTypes.isArrayOfType(validateTypes.isString),
        autoCreateTopics: validateTypes.isBoolean,
        portMapping: validateTypes.isObjectWithValuesOfType(validateTypes.isString),
        logLevel: validateTypes.isOneOf(values(LOG_LEVEL)),
      }

      const failures = validateTypes(validationSchema, config)

      expect(failures).toEqual([])
    })
  })

  describe('sad', () => {
    it(`should return a failure right away if there's no config`, () => {
      const failures = validateTypes({}, undefined)

      expect(failures.length).toEqual(1)
      expect(failures[0]).toMatch('Missing config to validate')
    })

    it(`should return a failure for invalid config`, () => {
      config.service = 1
      const validationSchema = {
        service: validateTypes.isString,
      }

      const failures = validateTypes(validationSchema, config)

      expect(failures.length).toEqual(1)
      expect(failures[0]).toMatch('service: Expected string | Received 1 (number)')
    })
  })
})
