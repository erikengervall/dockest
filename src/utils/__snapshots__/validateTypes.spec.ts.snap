// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateTypes happy should work for valid config 1`] = `Array []`;

exports[`validateTypes sad should return a failure for invalid config 1`] = `1`;

exports[`validateTypes sad should return a failure for invalid config 2`] = `"service: Expected string | Received 1 (number)"`;
